### PR TITLE
Refactor deprecated lifecycle methods in Table

### PR DESCRIPTION
**Pull request by kristofdombi**
### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / document update
- [ ] Component style update
- [ ] TypeScript definition update
- [X] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link

https://github.com/ant-design/ant-design/issues/9792

### 💡 Background and solution

Updated to `react` v16.9 and it started to log warnings about the unsafe lifecycle methods.

<details>
<summary>📸 Screenshot</summary>

image

</details>

I ran `npx react-codemod rename-unsafe-lifecycles`, following the React [blog post](https://reactjs.org/blog/2019/08/08/react-v16.9.0.html#renaming-unsafe-lifecycle-methods).

### 📝 Changelog

No more warnings about `componentWillReceiveProps` from `Table` and `SelectionCheckboxAll` components. 🙅‍♂ 

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |     -      |
| 🇨🇳 Chinese |     -      |

### ☑️ Self Check before Merge

- [X] Doc is updated/provided or not needed
- [X] Demo is updated/provided or not needed
- [X] TypeScript definition is updated/provided or not needed
- [X] Changelog is provided or not needed

### CODE REVIEW

**Comment by zombieJ** [comment]
Could you help to change to `getDerivedStateFromProps` instead of `UNSAFE_`?

**Comment by kristof0425** [comment]
This is all I could help at the moment. 🙂

**Comment by afc163** [comment]
oops, conflicted

**Review by afc163 — CHANGES_REQUESTED** [review]
conflicted

**Comment by afc163** [comment]
@kristof0425  ping

**Comment by kristof0425** [comment]
@afc163 I ran into an issue and I'm stuck with it. Let me describe it:

I put all the checks from `cWRP` to `cDM`, which contains a call to `this.store` as in `static getDerivedStateFromProps` I can't reach `this`. This solved some of the failing tests, but the rest of the failing tests are because of this. 
I suppose we'd need to update before render (putting everything to `getDerivedStateFromProps`)

I need some guidance here... 🤷‍♂ 

I had the idea, that we could create a HOC with the store and Table would get the `store.setState` as a `prop`, so we could reach it within the `static getDerivedStateFromProps` method and do the checks there.

**Comment by kristof0425** [comment]
I'm working on `Table.filter.test.js` now...

**Comment by kristof0425** [comment]
@afc163 Now all the tests pass except the `demo/dynamic-settings.md`. Could you help there?
It doesn't seem to be a Table problem.

**Comment by zombieJ** [comment]
You can check CI for the error info.

**Comment by kristof0425** [comment]
I can't replicate the failing tests in `*_15` environments.

Could you help?

**Comment by afc163** [comment]
Run this in local

```
npm i --no-save react@15 react-dom@15 react-test-renderer@15 enzyme-adapter-react-15
npm test
```

https://github.com/ant-design/ant-design/blob/master/scripts/install-react.sh

File: components/table/Table.tsx [hunk 1]
```diff
@@ -124,22 +184,83 @@ export default class Table<T> extends React.Component<TableProps<T>, TableState<
     childrenColumnName: 'children',
   };
 
-  CheckboxPropsCache: {
-    [key: string]: any;
-  };
+  static getDerivedStateFromProps(nextProps: TableProps<any>, prevState: TableState<any>) {
```
Comment (afc163) [inline components/table/Table.tsx]: You need `react-lifecycles-compat` to polyfill Table for react 15.

Comment (kristof0425) [reply components/table/Table.tsx]: Yes, I figured, but the tests are failing with that in place as well. 😞

**Comment by afc163** [comment]
1. You need react-lifecycles-compat to polyfill Table for react 15, like other components did.
2. Please send this PR to feature branch.

**Comment by 18601673727** [comment]
very much need this PR merged, thanks for all your efforts!

**Comment by afc163** [comment]
Merged in #18953

**Comment by deadcoder0904** [comment]
@afc163 when will this be released?

**Pull request closed without merging**